src/lib.rs

```diff
@@ -71,6 +71,14 @@
 #![deny(warnings, missing_docs, missing_debug_implementations, rust_2018_idioms)]
 #![doc(html_root_url = "https://docs.rs/bytes/0.5.0")]
 
+#![no_std]
+
+#[cfg(feature = "std")]
+extern crate std;
+
+#[cfg(not(feature = "std"))]
+extern crate alloc;
```

Comment:
If this is targeting 1.36+, you can just use `extern crate alloc` and be done with it. Either way you're linking `liballoc`, it's just in one case you're going through the `std` facade.

Also note in #255 I had some of the core functionality of `bytes` available in a no-`alloc` context, which could be useful in conjunction with crates like [`heapless`](https://docs.rs/heapless/latest/heapless/)

Reply:
> Also note in #255 I had some of the core functionality of bytes available in a no-alloc context, which could be useful in conjunction with crates like heapless

I didn't understand this, doesn't bytes rely pretty fundamentally on `vec`? What parts of this crate were you going to build without pulling in `vec`? Just the `buf` and `buf_mut` parts?

What would you think about putting these trait objects in a smaller crate that would not have `alloc` dependency in any configuration? That's my instinct anyways

Reply:
Hmm maybe that smaller crate would still need some flags I guess, on looking at `buf_impl.rs` again

Reply:
@tarcieri yeah i guess i see what you are saying

do you want to see an `alloc` feature as well here?

Reply:
The idea would be to use e.g. `heapless::Vec` in lieu of `alloc::vec::Vec. Probably a distraction to bring them up in this PR, sorry. Would be good to land this and circle back on full `no_std` support.

Reply:
I'd say you don't need to add an `alloc` feature until there's a useful `no_std` mode without it, and that's probably best saved for a separate PR dealing with it specifically.

Reply:
if we ultimately are going to want an alloc feature then it would be nice to figure that out beforehand because at this revision, `default-features = false` is the alloc configuration, so it would be a breaking change for anyone with that in their Cargo.toml if we add alloc feature later, and `default-features=false` becomes core-only

Reply:
imo it might be nice to put the `Buf` traits in their own crate if core-only / heapless is a goal, that would reduce the scope of bytes crate to, providing the wrapper around `Vec` that implements those traits, and reduces the configuration surface area of `bytes` crate

Reply:
If you want to tackle it now, I'm not going to complain 😉

A crate containing only the core traits sounds good to me

Reply:
(sorry for confusion, my previous 4 comments were written without any of your comments being delivered from some reason)

i think i will play around with the smaller crate idea later, but i'm gonna hold off on it, and an alloc feature, for this PR